sensors: Add channel specifiers

teburd opened the pull request:
Use a structured channel specifier rather than a single enum when specifying channels to read in the new read/decoder API.

Implements a slight variation of #63830

yperess reviewed (dismissed):
Seems reasonable to me, please run clang-format and push.

tristan-google commented:
If it's not too much trouble it'd be nice to separate the clang-formatting into a different commit. The new channel struct looks good to me.

teburd commented:
> If it's not too much trouble it'd be nice to separate the clang-formatting into a different commit. The new channel struct looks good to me.

yeah of course, I can split the commits or drop the unchanged areas that clang-format touched as a whole... seems clang-format version dependent on what it does sometimes

Commit:
sensors: Add channel specifier

Use a structured channel specifier rather than a single enum when
specifying channels to read in the new read/decoder API.

Replaces usages of a seperate channel and channel_index parameter
where previously used with a struct sensor_chan_spec.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Has turned into a bit of a minor rabbit hole and might not finish before ZDS at this point

teburd commented:
Updated the emulated device implementations to all use sensor_chan_spec which should fix the remaining build issues hopefully

teburd commented:
Update #2:

Fixes remaining issues with tests not being updated to use chan_spec.

At some point attr_set/attr_get should be updated but that I think is another PR.

teburd commented:
Fixed an issue with the bmi160 fixture and CI is finally green, please take a look now @yperess @tristan-google @laurenmurphyx64 @MaureenHelm, thanks!

str4t0m commented on tests/drivers/build_all/sensor/src/generic_test.c:
```diff
@@ -155,16 +156,17 @@ static void run_generic_test(const struct device *dev)
 
 		/* Set this iteration's expected values in emul for every supported channel */
 		for (size_t i = 0; i < iodev_read_config.count; i++) {
-			enum sensor_channel ch = iodev_all_channels[i];
+			struct sensor_chan_spec ch = iodev_all_channels[i];
```
NIT consider changing the name to something like `ch_spec`, as there is already a `ch` variable of a different type(`enum sensor_channel`) in same function above(line 106).

str4t0m reviewed (changes requested):
Looks good, found a few small things.

str4t0m reviewed (changes requested):
I've just noticed that the Bosch bma4xx also implements the emulator api, so the `xxx_get_sample_range` and `xxx_set_channel` should be changed there as well.

ubieda commented on drivers/sensor/sensor_shell.c:
```diff
@@ -360,12 +362,10 @@ void sensor_shell_processing_callback(int result, uint8_t *buf, uint32_t buf_len
 			continue;
 		}
 
-		while (decoder->get_frame_count(buf, channel, channel_idx, &frame_count) == 0) {
```
Since we're no longer using `channel_idx`, let's then increase `chan_spec.chan_idx` on every loop iteration (Otherwise, issuing `sensor get <sensor-id>` results in an endless loop of `chan_idx = 0` readings). The suggested fix:

``` c

	for (int channel = 0; channel < SENSOR_CHAN_ALL; ++channel) {
		...
		while (decoder->get_frame_count(buf, chan_spec, &frame_count) == 0) {
			...
			++chan_spec.chan_idx;
		}

```

I'd even go further and suggest scrubbing this local variable entirely as it's no longer needed.

ubieda replied:
Suggestion brought up on the Sensor WG Meeting: We probably want to add a testsuite verifying the sensor shell commands are kept functional across changes, in a generalized fashion, as the generic test covers all sensors.

(https://github.com/zephyrproject-rtos/zephyr/pull/72052 adds a test-suite for the proximity sensor shell commands)

In any case, not considered in scope for this PR. Just capturing it so we don't forget.

teburd commented:
WG May 6, 2024 (Summarized)

* @MaureenHelm noted chan specs are passed by value, and some comments around this are worth adding as to why
* @ubieda noted the shell broke, and tests validating shell commands would help, #72052 adds the start of this)

teburd commented:
Update 4: Fix sensor shell build issues

ubieda reviewed (dismissed):
I still see the issue with the Sensor Shell command, however, it's easy to fix and there are many PRs depending on this one. I'm approving it since I'm good with merging in and follow-up with subsequent PRs to complement it.

PR with Shell fix: #72542

MaureenHelm commented on drivers/sensor/bosch/bma4xx/bma4xx.c:
```diff
@@ -34,19 +34,19 @@ static int bma4xx_offset_to_reg_val(const struct sensor_value *val, uint8_t *reg
  * @brief Set the X, Y, or Z axis offsets.
  */
-static int bma4xx_attr_set_offset(const struct device *dev, enum sensor_channel chan,
+static int bma4xx_attr_set_offset(const struct device *dev, enum sensor_channel ch,
```
Should this be:
```suggestion
static int bma4xx_attr_set_offset(const struct device *dev, struct sensor_chan_spec ch,
```
?

teburd replied:
Yes it should be, thank you

teburd replied:
In looking at this closer it seems like its correct as it is, the set_offset function here is being called by bma4xx_attr_set. I didn't update the sensor_attr_get/sensor_attr_set functions to take a sensor_chan_spec as this would be a breaking change.

teburd replied:
Agreed, change shouldn't have made it in, removed this change from the PR

MaureenHelm commented on drivers/sensor/bosch/bma4xx/bma4xx.c:
```diff
@@ -202,7 +202,7 @@ static int bma4xx_attr_set_bwp(const struct device *dev, const struct sensor_val
  * @brief Implement the sensor API attribute set method.
  */
-static int bma4xx_attr_set(const struct device *dev, enum sensor_channel chan,
+static int bma4xx_attr_set(const struct device *dev, enum sensor_channel ch,
```
Same here

teburd replied:
attr set and attr get are not covered by this PR as that would be a breaking API change, but if the name of the variable needs changing I'll take a closer look at. thank you for spotting these things!

teburd replied:
Agreed, rename should not have occurred, undone

str4t0m reviewed (dismissed):
My concerns have been addressed, thanks!

It may be best to include the sensor shell fix from 72542 here already, but leaving that to the sensor maintainers.

teburd commented:
Update 5: Revert some inadvertent variable renames that had nothing to do with the channel spec changes
